### PR TITLE
gtk: add option to not link against libX11

### PR DESCRIPTION
**Pull request by jcollie**
Possible fix for #3477. Needs testing.

### CODE REVIEW

**Comment by tristan957** [comment]
Does the zig build system have a notion of Meson's feature options where you can detect if GTK was built without X11, and then build ghostty without X11 support too?

File: build.zig [hunk 1]
```diff
@@ -105,6 +105,12 @@ pub fn build(b: *std.Build) !void {
         "Enables the use of Adwaita when using the GTK rendering backend.",
     ) orelse true;
 
+    config.x11 = b.option(
+        bool,
+        "gtk-x11",
+        "Enables the linking against X11 libraries when using the GTK rendering backend.",
```
Comment (tristan957) [inline build.zig:111]:
```suggestion
        "Enables linking against X11 libraries when using the GTK rendering backend.",
```

File: src/apprt/gtk/ghostty_gtk_compat.h [hunk 2]
```diff
@@ -12,3 +12,11 @@
...
 #define G_APPLICATION_DEFAULT_FLAGS G_APPLICATION_FLAGS_NONE
 #endif
+
+#ifndef False
+#define False 0
+#endif
+
+#ifndef True
+#define True 1
+#endif
```
Comment (tristan957) [inline src/apprt/gtk/ghostty_gtk_compat.h:22]: Why is this necessary?

Comment (jcollie) [reply src/apprt/gtk/ghostty_gtk_compat.h:22]: Somewhere in `<gdk/x11/gdkx.h>` or `<X11/XKBlib.h>` C macros for `True` and `False` get defined and they are used in a couple of places in `gtk/Window.zig` and `gtk/Surface.zig`. I suppose it would be just as easy to replace the uses in the Zig code.

**Comment by jcollie** [comment]
> Does the zig build system have a notion of Meson's feature options where you can detect if GTK was built without X11, and then build ghostty without X11 support too? Essentially, just automatically do the right thing.

If there's a command that we can run to determine if GTK was built without X11, we should be able to take advantage of that in `build.zig`. I'm not aware of a native feature in Zig that does that.

File: src/apprt/gtk/ghostty_gtk_compat.h (hunk 2 above)
Comment (tristan957) [reply src/apprt/gtk/ghostty_gtk_compat.h:22]: Yeah I would recommend that

**Comment by tristan957** [comment]
```
$ pkg-config --variable=targets gtk4
broadway wayland x11
```

**Comment by tristan957** [comment]
`pkgconf` is also another tool for querying, so you may want to check both. Not sure if `pkg-config` is guaranteed to exist.

**Comment by mitchellh** [comment]
Ghostty already requires pkg-config to build so it’s fine to rely on it.

**Comment by jcollie** [comment]
> `pkgconf` is also another tool for querying, so you may want to check both. Not sure if `pkg-config` is guaranteed to exist.

Pretty sure that the Zig build system requires `pkg-config` to exist to find system libraries, so I think that we can depend on it being available.

**Comment by tristan957** [comment]
Even better!

**Comment by jcollie** [comment]
> ```
> $ pkg-config --variable=targets gtk4
> broadway wayland x11
> ```

That should be easy enough to check for.

**Comment by jcollie** [comment]
OK, the latest commit _should_ detect X11 support in GTK4 but I don't have a system handy that has GTK4 without X11 support. I _thought_ it would be possible to set that up in the devshell but I wasn't able to get that to work.

So if someone has a system without X11 installed it would be great if they can test it out.

**Comment by mitchellh** [comment]
This looks good to me. I want to make a couple stylistic tweaks but this is well done!

**Comment by mitchellh** [comment]
Moved the GTK tests to a matrix. If that's good will merge.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -329,25 +329,53 @@ jobs:
...
       - name: Test System Build
         run: nix develop -c zig build --system ${ZIG_GLOBAL_CACHE_DIR}/p
 
+  test-gtk:
+    strategy:
+      fail-fast: false
+      matrix:
+        adwaita: ["true", "false"]
+        x11: ["true", "false"]
+    runs-on: namespace-profile-ghostty-md
```
Comment (tristan957) [inline .github/workflows/test.yml]: I would suggest using enum like strings. Otherwise you won't know which job failed because it will look like `test-gtk(true, false)` or something

Comment (mitchellh) [reply .github/workflows/test.yml]: I just named them using the `name:` key. Is there some other way? That's what I'm used to doing.

Comment (tristan957) [reply .github/workflows/test.yml]: It is just a UX problem in my opinion of GitHub that have to be worked around. No biggie
